File: BilisPera/BilisPera/Utils/BPADTools.m
//
//  BPADTools.m
//  BilisPera
//
//  Created by BHJ on 2025/8/25.
//

#import "BPADTools.h"
#import <AdSupport/AdSupport.h>
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import "BPKeyChainWrapper.h"

NS_ASSUME_NONNULL_BEGIN

@interface BPADTools ()

@property (nonatomic, strong) NSString *idfvString;

@end

@implementation BPADTools

+ (instancetype)shared {
    static BPADTools *_shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _shared = [[BPADTools alloc] init];
    });
    return _shared;
}

- (void)registerIDFAAndTrack {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        __weak typeof(self) weakSelf = self;
        [self registerIDFAWithCompletion:^(BOOL success) {
            if (!success) return;
            if (weakSelf.trackCount < 2) {
                [[TrackTools shared] trackForGoogleMarket];
                weakSelf.trackCount += 1;
            }
        }];
    });
}

-(void)registerIDFAWithCompletion:(simpleBoolCompletion)completion{
    [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
        if (status == ATTrackingManagerAuthorizationStatusNotDetermined) {
            [self registerIDFAWithCompletion:completion];
        }else if(status == ATTrackingManagerAuthorizationStatusAuthorized){
            self.idfvString = [[[UIDevice currentDevice] identifierForVendor] UUIDString] ?: @"";
            NSData *idfvData = [self.idfvString dataUsingEncoding:NSUTF8StringEncoding];
            if (idfvData) {
                [BPKeyChainWrapper saveData:idfvData forKey:@"IDFV"];
            }
            completion(YES);
        }else{
            completion(NO);
        }
    }];
}


-(NSString *)getIDFV{
    NSData *data = [BPKeyChainWrapper loadDataForKey:@"IDFV"];
    if (data) {
        NSString *idfvStr = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        return idfvStr ?: @"";
    }
    return @"";
}

-(NSString *)getIDFA{
    return [[[ASIdentifierManager sharedManager] advertisingIdentifier] UUIDString] ?: @"";
}

@end

NS_ASSUME_NONNULL_END

File: BilisPera/BilisPera/Utils/BPADTools.swift
//
//  BPADTools.swift
//  BilisPera
//
//  Created by BHJ on 2025/8/22.
//

import Foundation
import AdSupport
import AppTrackingTransparency
import UIKit

@objc
@MainActor
class BPADTools: NSObject {
   @objc static let shared = BPADTools()
   private let idfvString: String = UIDevice.current.identifierForVendor?.uuidString ?? ""
    @objc var trackCount: Int = 0
    
   @objc func registerIDFAAndTrack(){
        Task{
            let delay = UInt64(1.5) * 1_000_000_000
            try await Task.sleep(nanoseconds: delay)
            defer{
                if self.trackCount < 2{
                    TrackTools.shared().trackForGoogleMarket()
                    self.trackCount += 1
                }
            }
            guard await registerIDFA() else {return}
        }
    }
    
  @objc  func registerIDFA() async -> Bool{
        let status =  await ATTrackingManager.requestTrackingAuthorization()
        switch status {
        case .authorized:
            // 用户已授权，可以访问IDFA
            let idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
            if let idfaData = idfa.data(using: .utf8) {
                _ = BPKeyChainWrapper.save(key: "IDFA", data: idfaData)
            }
            if let idfvData = self.idfvString.data(using: .utf8){
                _ = BPKeyChainWrapper.save(key: "IDFV", data: idfvData)
            }
            return true
        case .notDetermined:
            return await registerIDFA()
        case .denied, .restricted:
            return false
        @unknown default:
            return false
        }
    }
    
  @objc func getIDFV() -> String{
        guard let data = BPKeyChainWrapper.load(key: "IDFV"),let string = String(data: data, encoding: .utf8) else { return ""}
        return string
    }
    
   @objc func getIDFA() -> String{
      return  ASIdentifierManager.shared().advertisingIdentifier.uuidString
    }
}
